src/mlx5.cpp

```diff
@@ -330,6 +318,225 @@ int gds_mlx5_get_dword_wait_info(uint32_t *ptr, uint32_t value, int flags, gds_m
+                seg = (void *)((char *)qp->dv_qp.sq.buf + (idx * qp->dv_qp.sq.stride));
+
+                ctrl_seg = (struct mlx5_wqe_ctrl_seg *)seg;
+                size = sizeof(struct mlx5_wqe_ctrl_seg) / 16;
+                seg = (void *)((char *)seg + sizeof(struct mlx5_wqe_ctrl_seg));
+
+                switch (qp->qp->qp_type) {
+                        case IBV_QPT_UD:
+                                ret = set_datagram_seg((struct mlx5_wqe_datagram_seg *)seg, p_ewr);
+                                if (ret)
+                                        goto out;
+                                size = sizeof(struct mlx5_wqe_datagram_seg) / 16;
+                                seg = (void *)((char *)seg + sizeof(struct mlx5_wqe_datagram_seg));
+                                if (seg == qend)
+                                        seg = qp->dv_qp.sq.buf;
+                                break;
+                        case IBV_QPT_RC:
+                                break;
+                        default:
+                                gds_err("Encountered unsupported qp_type. We currently support only IBV_QPT_UD\n");
+                                ret = EINVAL;
+                                goto out;
+                }
+
+                data_seg = (struct mlx5_wqe_data_seg *)seg;
+                for (i = 0; i < p_ewr->num_sge; ++i) {
+                        if (data_seg == qend) {
+                                seg = qp->dv_qp.sq.buf;
+                                data_seg = (struct mlx5_wqe_data_seg *)seg;
+                        }
+                        if (p_ewr->sg_list[i].length) {
+                                mlx5dv_set_data_seg(data_seg, p_ewr->sg_list[i].length, p_ewr->sg_list[i].lkey, p_ewr->sg_list[i].addr);
+                                ++data_seg;
+                                size += sizeof(struct mlx5_wqe_data_seg) / 16;
+                        }
+                }
+
+                mlx5dv_set_ctrl_seg(ctrl_seg, qp->sq_cur_post & 0xffff, MLX5_OPCODE_SEND, 0, qp->qp->qp_num, MLX5_WQE_CTRL_CQ_UPDATE, size, 0, 0);
+
+                qp->send_cq.wrid[idx] = p_ewr->wr_id;
+                qp->sq_cur_post += (size * 16 + qp->dv_qp.sq.stride - 1) / qp->dv_qp.sq.stride;
+        }
+
+        commit->rollback_id = qp->peer_scur_post | ((uint64_t)qp->sq_cur_post << 32);
+        qp->peer_scur_post = qp->sq_cur_post;
+
+        wr = commit->storage;
+
+        wr->type = GDS_MLX5_PEER_OP_STORE_DWORD;
+        wr->wr.dword_va.data = htonl(qp->sq_cur_post & 0xffff);
+        wr->wr.dword_va.target_id = qp->peer_va_id_dbr;
+        wr->wr.dword_va.offset = sizeof(uint32_t) * MLX5_SND_DBR;
+        wr = wr->next;
+
+        wr->type = GDS_MLX5_PEER_OP_FENCE;
+        wr->wr.fence.fence_flags = GDS_PEER_FENCE_OP_WRITE | GDS_PEER_FENCE_FROM_HCA | GDS_PEER_FENCE_MEM_SYS;
+        wr = wr->next;
+
+        wr->type = GDS_MLX5_PEER_OP_STORE_QWORD;
```

Comment:
we have extra fixes and optimizations in https://github.com/gpudirect/libmlx5/commits/fixes and https://github.com/gpudirect/libmlx5/commits/copyblock-4.3 which should be pulled here.
would you mind filing a new issue?

Reply:
Filed an issue.